### PR TITLE
feat(daily): add index to alert_log that speeds up daily cleanup process

### PR DESCRIPTION
**Pull request by rkojedzinszky**
Add index to speed up daily cleanup process.

In our deployment, running [this](https://github.com/librenms/librenms/blob/20e88e6cf4e9df1398be456a16df9c6349156311/daily.php#L263) as a SELECT takes much time:

```
mysql> select count(*) FROM alert_log WHERE id IN ( SELECT id FROM ( SELECT id FROM alert_log a1 WHERE time_logged < DATE_SUB(NOW(),INTERVAL 365 DAY) AND (device_id, rule_id, time_logged) NOT IN ( SELECT device_id, rule_id, max(time_logged) FROM alert_log a2 WHERE a1.device_id = a2.device_id AND a1.rule_id = a2.rule_id AND a2.time_logged < DATE_SUB(NOW(),INTERVAL 365 DAY) ) ) as c);
+----------+
| count(*) |
+----------+
|      242 |
+----------+
1 row in set (10 min 17.58 sec)

```

With the index:
```
mysql> select count(*) FROM alert_log WHERE id IN ( SELECT id FROM ( SELECT id FROM alert_log a1 WHERE time_logged < DATE_SUB(NOW(),INTERVAL 365 DAY) AND (device_id, rule_id, time_logged) NOT IN ( SELECT device_id, rule_id, max(time_logged) FROM alert_log a2 WHERE a1.device_id = a2.device_id AND a1.rule_id = a2.rule_id AND a2.time_logged < DATE_SUB(NOW(),INTERVAL 365 DAY) ) ) as c);
+----------+
| count(*) |
+----------+
|      250 |
+----------+
1 row in set (1.10 sec)

```

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [ ] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [ ] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by rkojedzinszky** [comment]
I am still evaluating if the order of columns could be changed, and if existing indexes could be dropped with this added or not.

**Comment by murrant** [comment]
How do these index changes affect other queries?

**Comment by rkojedzinszky** [comment]
> How do these index changes affect other queries?

Should not affect.

The new index `alert_log_device_id_rule_id_time_logged_index` with columns `[device_id, rule_id, time_logged] ` can be used for queries filtering for equality on the leftmost columns, so an index on `[device_id]` or `[device_id, rule_id]` becomes redundant.

Also, the index `alert_log_rule_id_device_id_state_index` with columns `[rule_id, device_id, state]` can be used for queries with equality filtering for equality on the leftmost columns, so an index on `[rule_id]` or `[rule_id, device_id]` becomes redundant.

In other words:
- `alert_log_device_id_rule_id_time_logged_index` will be used wherever  `alert_log_device_id_index` was used
- `alert_log_rule_id_device_id_state_index` will be used wherever `alert_log_rule_id_device_id_index` or `alert_log_rule_id_index` was used

**Comment by murrant** [comment]
From my experience it is possible to query rule_id (without device_id) and miss the index.  But I could be wrong and that could be a special case I'm not exactly a DB expert :D

**Comment by murrant** [comment]
Oh, I didn't look at the existing indexes that weren't removed, yeah, those indexes should cover it :D

**Review by murrant — APPROVED** [review]

**Comment by murrant** [comment]
P.S. the query you are optimizing is a little crazy :D

**Comment by rkojedzinszky** [comment]
> P.S. the query you are optimizing is a little crazy :D

@murrant thanks for the merge. Can you please take a look at https://github.com/librenms/librenms/pull/16763 too? :)
